biscuit/src/Auth/Biscuit/Datalog/Parser.hs

```diff
@@ -1,448 +1,393 @@
+forbid mkError p = do
+  (sp, _) <- getSpan p
+  registerError mkError sp
+
+variableParser :: Parser Text
+variableParser =
+  C.char '$' *> takeWhile1P (Just "_, :, or any alphanumeric char") (\c -> c == '_' || c == ':' || isAlphaNum c)
 
-import           Auth.Biscuit.Crypto            (PublicKey,
-                                                 readEd25519PublicKey)
-import           Auth.Biscuit.Datalog.AST
-import qualified Data.Text as Text
-
-class ConditionalParse a v where
-  ifPresent :: String -> Parser a -> Parser v
-
-instance ConditionalParse a Void where
-  ifPresent name _ = fail $ name <> " is not available in this context"
-
-instance ConditionalParse m m where
-  ifPresent _ p = p
-
-class SetParser (inSet :: IsWithinSet) (ctx :: DatalogContext) where
-  parseSet :: Parser (SetType inSet ctx)
-
-instance SetParser 'WithinSet ctx where
-  parseSet = fail "nested sets are forbidden"
-
-instance SetParser 'NotWithinSet 'WithSlices where
-  parseSet = Set.fromList <$> (char '[' *> commaList0 termParser <* char ']')
-
-instance SetParser 'NotWithinSet 'Representation where
-  parseSet = Set.fromList <$> (char '[' *> commaList0 termParser <* char ']')
-
-class ScopeParser (evalCtx :: EvaluationContext) (ctx :: DatalogContext) where
-  parseBlockId :: Parser (BlockIdType evalCtx ctx)
-
-instance ScopeParser 'Repr 'Representation where
-  parseBlockId = publicKeyParser
+haskellVariableParser :: Parser Text
+haskellVariableParser = l $ do
+  _ <- chunk "${"
+  leadingUS <- optional $ C.char '_'
+  x <- if isJust leadingUS then C.letterChar else C.lowerChar
+  xs <- takeWhileP (Just "_, ', or any alphanumeric char") (\c -> c == '_' || c == '\'' || isAlphaNum c)
+  _ <- C.char '}'
+  pure . maybe id T.cons leadingUS $ T.cons x xs
+
+setParser :: Parser (Set (Term' 'WithinSet 'InFact 'WithSlices))
+setParser = do
+  _ <- l $ C.char '['
+  ts <- sepBy (termParser (forbid VarInSet variableParser) (forbid NestedSet setParser)) (l $ C.char ',')
+  _ <- l $ C.char ']'
+  pure $ Set.fromList ts
+
+factTermParser :: Parser (Term' 'NotWithinSet 'InFact 'WithSlices)
+factTermParser = termParser (forbid VarInFact variableParser)
+                            setParser
+
+predicateTermParser :: Parser (Term' 'NotWithinSet 'InPredicate 'WithSlices)
+predicateTermParser = termParser variableParser
+                                 setParser
+
+termParser :: Parser (VariableType inSet pof)
+           -> Parser (SetType inSet 'WithSlices)
+           -> Parser (Term' inSet pof 'WithSlices)
+termParser parseVar parseSet = l $ choice
+  [ Antiquote . Slice <$> haskellVariableParser
+  , Variable <$> parseVar
+  , TermSet <$> parseSet
+  , LBytes <$> (chunk "hex:" *> hexParser)
+  , LDate <$> try rfc3339DateParser
+  , LInteger <$> L.signed C.space L.decimal
+  , LString . T.pack <$> (C.char '"' *> manyTill L.charLiteral (C.char '"'))
+  , LBool <$> choice [ True <$ chunk "true"
+                     , False <$ chunk "false"
+                     ]
+  ]
```

Comment:
Have you tested it with funny cases like : variable prefixed by `hex` ?

1. Shouldn't `LBytes` be prioritized other `Variable`
2. I'm not sure if megaparsec backtracks by default in a choice, I used to do it explicitly with parsec. Is that needed here?

Reply:
> Shouldn't LBytes and LBool be prioritized other Variable?

I guess not, variables start with a dollar sign.

Reply:
> (does) megaparsec backtracks by default in a choice?

No it doesn't, it's the purpose of `try` you use elsewhere (like in parsec) and I guess it's not required here as there's no ambiguity in this subsyntax.

Reply:
antiquote, variable, term and bytes all start with `chunk`, which backtracks.

The only ambiguity is between integers and dates, hence the explicit `try`